gnovm/cmd/gno/fmt.go

```diff
@@ -0,0 +1,307 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"go/format"
+	"go/parser"
+	"go/scanner"
+	"go/token"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/gnolang/gno/gnovm/pkg/gnoenv"
+	"github.com/gnolang/gno/gnovm/pkg/gnoimports"
+	"github.com/gnolang/gno/tm2/pkg/commands"
+	"github.com/rogpeppe/go-internal/diff"
+)
+
+type fmtCfg struct {
+	write   bool
+	quiet   bool
+	diff    bool
+	verbose bool
+	imports bool
+	strict  bool
+	include fmtIncludes
+}
+
+var defaultFmtOptions = &fmtCfg{
+	imports: true,
+}
+
+func newFmtCmd(io commands.IO) *commands.Command {
+	cfg := &fmtCfg{}
+	return commands.NewCommand(
+		commands.Metadata{
+			Name:       "fmt",
+			ShortUsage: "gno fmt [flags] [path ...]",
+			ShortHelp:  "Run gno file formatter.",
+			LongHelp:   "The `gno fmt` tool processes, formats, and cleans up `gno` source files.",
+		},
+		cfg,
+		func(_ context.Context, args []string) error {
+			return execFmt(cfg, args, io)
+		})
+}
+
+func (c *fmtCfg) RegisterFlags(fs *flag.FlagSet) {
+	fs.BoolVar(
+		&c.write,
+		"w",
+		defaultFmtOptions.write,
+		"write result to (source) file instead of stdout",
+	)
+
+	fs.BoolVar(
+		&c.verbose,
+		"v",
+		defaultFmtOptions.verbose,
+		"verbose mode",
+	)
+
+	fs.Var(
+		&c.include,
+		"i",
+		"specify additional directories containing packages to resolve",
+	)
+
+	fs.BoolVar(
+		&c.quiet,
+		"q",
+		defaultFmtOptions.verbose,
+		"quiet mode",
+	)
+
+	fs.BoolVar(
+		&c.strict,
+		"strict",
+		defaultFmtOptions.strict,
+		"when enabled, fail if any parse errors are encountered",
+	)
+
+	fs.BoolVar(
+		&c.imports,
+		"imports",
+		defaultFmtOptions.imports,
+		"attempt to format, resolve and sort file imports",
+	)
+
+	fs.BoolVar(
+		&c.diff,
+		"diff",
+		defaultFmtOptions.diff,
+		"print and make the command fail if any diff is found and write is disabled",
+	)
+}
+
+type fmtProcessFileFunc func(file string, io commands.IO) []byte
+
+func execFmt(cfg *fmtCfg, args []string, io commands.IO) error {
+	cfg.write = !cfg.diff && cfg.write
```

Comment:
beside the readonly, I was thinking that it can be confusing to have the diff and write flag at the same time, but after a second thought, I think it's fine